rfd/0010-api.md

```diff
@@ -0,0 +1,235 @@
+*  It's impossible to generate library in any other language other than Go due to unclear API surface.
+*  Some client logic is residing in [plugins](https://github.com/gravitational/teleport-plugins/blob/master/access/access.go), while other is in `lib/auth`.
+* Code in `lib/auth` uses some concepts unfamiliar to Go users:
+
+```go
+// connectClient establishes a gRPC connection to an auth server.
+func connectClient() (*auth.Client, error) {
+    tlsConfig, err := LoadTLSConfig("certs/api-admin.crt", "certs/api-admin.key", "certs/api-admin.cas")
+    if err != nil {
+        return nil, fmt.Errorf("Failed to setup TLS config: %v", err)
+    }
+
+    // must prase addr list panics and is not necessary
+    authServerAddr := utils.MustParseAddrList("127.0.0.1:3025")
+    clientConfig := auth.ClientConfig{Addrs: authServerAddr, TLS: tlsConfig}
+
+    // TLS client is the only client
+    return auth.NewTLSClient(clientConfig)
+```
+
+## Details
+
+### Approach
+
+Keep Teleport's API layer unified, avoid having internal API vs external API layer.
+The access to the API surface is limited by role based access control.
+
+Avoid large refactors, instead use Go's type aliases to move things around while keeping things backwards compatible.
+
+```go
+// Preserve full backwards compatibility
+type services.Role = types.Role
+type auth.Service = proto.Service
+```
+
+Use Swagger and OpenAPI to document and create SDKs.
+
+#### Move protobuf types and services
+
+Point protobuf generators to generate types to another folder and use a system of submodules:
+
+```bash
+$ ls github.com/gravitational/teleport/api
+# root folder with API specs and is a submodule
+api/
+  # open API spec used for swagger, see for example K8s Swagger spec
+  # https://github.com/kubernetes/kubernetes/tree/master/api/openapi-spec
+  openapi/
+    spec.json
+    # move generated services types here, while keeping the original proto files
+    # in place in lib/services/services.proto
+  types/
+    types.pb.go
+    wrappers.pb.go
+  # generate auth service lib/auth/proto here
+  client/
+    authservice.pb.go
+```
+
+#### Move client implementation to `api/client`
+
+Move relevant lib/client.*Client code to `api/client`.
+Track dependencies, make sure that new `api/client` module has no dependencies
+on the outside modules anywhere in `lib`.
+
+**Track dependencies**
+
+Track dependencies to be limited to:
+  * `gravitational/roundrip`,  `gravitational/trace`
+  * grpc google package and its dependencies
+  * protobuf generated code
+  * Keep some auth service specific code like `utils.Addr` out of the `api/client`
+
+**Go native user experience**
+
+Achieve the user experience in Client:
+
+```go
+import (
+   "github.com/gravitational/teleport/api/client"
+)
+
+func main() {
+   // TLS client is the only TLS client supported in teleport
+   client, err := client.New(client.Config{
+     Addrs: []string{"localhost:3025"},
```

Comment:
This should transparently handle dialing via a proxy.
Please make that an explicit requirement in the RFD